build fail looks related
```
======================================================================
ERROR: test_case_detail_tabs_with_nodesets (corehq.apps.app_manager.tests.test_suite.SuiteTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_suite.py", line 342, in test_case_detail_tabs_with_nodesets
    self._test_generic_suite("app_case_detail_tabs_with_nodesets", 'suite-case-detail-tabs-with-nodesets')
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/util.py", line 71, in _test_generic_suite
    self.assertXmlEqual(self.get_xml(suite_tag), app.create_suite())
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/models.py", line 4712, in create_suite
    return SuiteGenerator(self).generate_suite()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/suite_xml/generator.py", line 72, in generate_suite
    EntryInstances(self.suite, self.app, self.modules).update_suite()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/suite_xml/post_process/instances.py", line 14, in update_suite
    self.add_referenced_instances(entry, details_by_id, relevance_by_menu, menu_by_command)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/suite_xml/post_process/instances.py", line 59, in add_referenced_instances
    instances = EntryInstances.get_required_instances(xpaths)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/suite_xml/post_process/instances.py", line 80, in get_required_instances
    raise UnknownInstanceError("Instance reference not recognized: {}".format(instance_name))
UnknownInstanceError: Instance reference not recognized: external
----------------------------------------------------------------------
```

--- a/corehq/apps/app_manager/suite_xml/post_process/instances.py
+++ b/corehq/apps/app_manager/suite_xml/post_process/instances.py
@@ -77,7 +77,9 @@ class EntryInstances(PostProcessor):
                 if instance:
                     instances.add(instance)
                 else:
-                    raise UnknownInstanceError("Instance reference not recognized: {}".format(instance_name))
+                    raise UnknownInstanceError(
+                        "Instance reference not recognized: {} in xpath \"{}\""
+                        .format(instance_name, xpath))
         return instances
 
 

--- a/corehq/apps/app_manager/suite_xml/post_process/instances.py
+++ b/corehq/apps/app_manager/suite_xml/post_process/instances.py
@@ -56,14 +56,15 @@ class EntryInstances(PostProcessor):
                     for datum in frame.datums:
                         xpaths.add(datum.value)
         xpaths.discard(None)
-        instances = EntryInstances.get_required_instances(xpaths)
+        instances, unknown_instance_ids = EntryInstances.get_required_instances(xpaths)
 
-        entry.require_instance(*instances)
+        entry.require_instances(instances=instances, instance_ids=unknown_instance_ids)
 
     @staticmethod
     def get_required_instances(xpaths):
         instance_re = r"""instance\(['"]([\w\-:]+)['"]\)"""
         instances = set()
+        unknown_instance_ids = set()
         for xpath in xpaths:
             instance_names = re.findall(instance_re, xpath)
             for instance_name in instance_names:
@@ -77,10 +78,12 @@ class EntryInstances(PostProcessor):
                 if instance:
                     instances.add(instance)
                 else:
-                    raise UnknownInstanceError(
-                        "Instance reference not recognized: {} in xpath \"{}\""
-                        .format(instance_name, xpath))
-        return instances
+                    class UnicodeWithContext(unicode):
+                        pass
+                    instance_name = UnicodeWithContext(instance_name)
+                    instance_name.xpath = xpath
+                    unknown_instance_ids.add(instance_name)
+        return instances, unknown_instance_ids
 
 
 def get_instance_factory(scheme):

--- a/corehq/apps/app_manager/suite_xml/sections/details.py
+++ b/corehq/apps/app_manager/suite_xml/sections/details.py
@@ -367,7 +367,8 @@ def get_instances_for_module(app, module, additional_xpaths=None):
         for detail_id in detail_ids:
             xpaths.update(details_by_id[detail_id].get_all_xpaths())
 
-        return EntryInstances.get_required_instances(xpaths)
+        instances, _ = EntryInstances.get_required_instances(xpaths)
+        return instances
 
 
 class CaseTileHelper(object):

--- a/corehq/apps/app_manager/suite_xml/sections/entries.py
+++ b/corehq/apps/app_manager/suite_xml/sections/entries.py
@@ -201,8 +201,7 @@ class EntriesHelper(object):
                 for tab in module.case_details.long.tabs:
                     connectors.update(tab.connectors)
         for id, src in connectors.iteritems():
-            entry.require_instance(Instance(id=id, src=src))
-
+            entry.require_instances([Instance(id=id, src=src)])
 
     @staticmethod
     def get_assertion(test, locale_id, locale_arguments=None):

--- a/corehq/apps/app_manager/suite_xml/xml_models.py
+++ b/corehq/apps/app_manager/suite_xml/xml_models.py
@@ -4,6 +4,7 @@ from eulxml.xmlmap import (
     load_xmlobject_from_string,
 )
 from lxml import etree
+from corehq.apps.app_manager.exceptions import UnknownInstanceError
 
 
 class XPathField(StringField):
@@ -373,7 +374,7 @@ class Entry(OrderedXmlObject, XmlObject):
 
     assertions = NodeListField('assertions/assert', Assertion)
 
-    def require_instance(self, *instances):
+    def require_instances(self, instances=(), instance_ids=()):
         used = {(instance.id, instance.src) for instance in self.instances}
         for instance in instances:
             if (instance.id, instance.src) not in used:
@@ -392,6 +393,15 @@ class Entry(OrderedXmlObject, XmlObject):
                     self.node.remove(instance_node)
                     self.node.insert(self.node.index(command_node) + 1,
                                      instance_node)
+        covered_ids = {instance_id for instance_id, _ in used}
+        for instance_id in instance_ids:
+            if instance_id not in covered_ids:
+                raise UnknownInstanceError(
+                    "Instance reference not recognized: {} in xpath \"{}\""
+                    # to get xpath context to show in this error message
+                    # make instance_id a unicode subclass with an xpath property
+                    .format(instance_id, getattr(instance_id, 'xpath', "(Xpath Unknown)")))
+
         sorted_instances = sorted(self.instances,
                                   key=lambda instance: instance.id)
         if sorted_instances != self.instances:

